File: AdaptiveFilters/python/tutorial_1/calculate_output.py
import numpy as np
import matplotlib.pyplot as plt

def calculate_output(S_hat, size_y, N_fft=256, frameshift=64):
    """
    CALCULATE_OUTPUT Calculate the output of the Wiener filter
    
    Parameters:
    S_hat       -- Spectrogram (complex-valued STFT output) of the Wiener filter
    size_y      -- Length of the original noisy signal (used to size the output)
    N_fft       -- FFT resolution (optional, default=256)
    frameshift  -- Hop size in samples between frames (optional, default=64)

    Returns:
    s_hat       -- Time-domain output of the Wiener filter
    """
    
    N_f, N_t = S_hat.shape

    # Initialize the output signal with zeros
    s_hat = np.zeros(size_y)

    print('AAA')
    return

    # Convert each frame back to time domain and perform overlap-add
    for k in range(N_t):
        # Mirror and conjugate for real IFFT
        spectrum = np.concatenate([
            S_hat[:, k],
            np.conj(S_hat[-2:0:-1, k])
        ])

        print(k)
        print(S_hat[:, k])
        print(np.conj(S_hat[-2:0:-1, k]))

        # IFFT and scaling
        frame = np.real(np.fft.ifft(spectrum)) * frameshift / (N_fft / 2)

        # Overlap-add to the output signal
        start = k * frameshift
        end = start + N_fft

        if end <= size_y:
            s_hat[start:end] += frame
        else:
            s_hat[start:size_y] += frame[:size_y - start]

    return s_hat

File: AdaptiveFilters/python/tutorial_1/calculate_output.m
%CALCULATE_OUTPUT Calculate the output of the Wiener filter
%
%   s_hat = calculate_output(S_hat,size_y) transforms the output of the 
%           Wiener filter in the time domain    
%
%   INPUT
%   S_hat   - Spectra(spectrogram) of the output of Wiener filter
%   size_y  - Dimensions of the noisy signal
%   N_fft   - FFT resolution (optional)
%   frameshift  - Number of samples between signal blocks. (optional)
%
%   OUTPUT
%   s_hat   - Output of Wiener filter in time domain
%
function s_hat = calculate_output(S_hat, size_y, N_fft, frameshift)

    % Set default values.
    if nargin < 3, N_fft = 256; end
    if nargin < 4, frameshift = 64; end

    [N_f, N_t] = size(S_hat);
    
    % Initialization of s_hat
    s_hat = zeros(size_y);
    
    % Transformation from frequency domain to time domain 
    for k=1:N_t
        frame = real(ifft([S_hat(:,k);conj(S_hat(end-1:-1:2,k))]))*frameshift/(N_fft/2);
        s_hat(1+(k-1)*frameshift:N_fft+(k-1)*frameshift) = s_hat(1+(k-1)*frameshift:N_fft+(k-1)*frameshift) ...
                                                         + frame;
    end
end